cluster/saltbase/salt/kube-addons/kube-addon-update.sh

```diff
@@ -0,0 +1,390 @@
+# 2. Services are treated as controllers, i.e. they will be udpated
+#    if service name is changed. the suffix after the last '-' is
+#    treated as verson. But we don't name services like this.
+# 3. Services will not be updated unless thir name is changed,
+#    but for services we acually want updates without name change.
+# 4. Json files are not handled at all. Currently addons must be
+#    in yaml files
+# 5. exit code is prabably not always correct (I haven't checked
+#    carefully if it works in 100% cases)
+# 6. There are no unittests
+# 7. object kind is checked using 'grep' it's possible that it doesn't work in
+#    all cases
+# 8. Will not work if the total length of paths to addons is greater than
+#    bash can handle. Probably it is not a problem: ARG_MAX=2097152 on GCE.
+#
+
+# cosmetic improvements to be done
+# 1. improve the log function; add timestamp, file name, etc.
+# 2. logging doesn't work from files that print things out.
+# 3. kubectl prints the output to stderr (the output should be captured and then
+#    logged)
+
+
+
+# global config
+KUBECTL=${TEST_KUBECTL:-/usr/local/bin/kubectl}   # substitute for tests
+NUM_TRIES_FOR_CREATE=${TEST_NUM_TRIES:-100}
+DELAY_AFTER_CREATE_ERROR_SEC=${TEST_DELAY_AFTER_ERROR_SEC:=10}
+NUM_TRIES_FOR_STOP=${TEST_NUM_TRIES:-100}
+DELAY_AFTER_STOP_ERROR_SEC=${TEST_DELAY_AFTER_ERROR_SEC:=10}
+
+if [[ ! -f $KUBECTL ]]; then
+    echo "ERROR: kubectl command ($KUBECTL) not found or is not executable" 1>&2
+    exit 1
+fi
+
+
+# remember that you can't log from functions that print some output (because
+# logs are also printed on stdout)
+# $1 level
+# $2 message
+function log() {
+  # manage log levels manually here
+
+  # add the timestamp if you find it useful
+  case $1 in
+    DB3 )
+#        echo "$1: $2"
+        ;;
+    DB2 )
+#        echo "$1: $2"
+        ;;
+    DBG )
+#        echo "$1: $2"
+        ;;
+    INFO )
+        echo "$1: $2"
+        ;;
+    WRN )
+        echo "$1: $2"
+        ;;
+    ERR )
+        echo "$1: $2"
+        ;;
+    * )
+        echo "INVALID_LOG_LEVEL $1: $2"
+        ;;
+  esac
+}
+
+
+
+# $1 addon path
+# $2 addon type (e.g. ReplicationController)
+# echoes the string with paths to files containing addon for the given type
+# works only for yaml files (!) (ignores json files)
+function get-addons-from-disk() {
+    # prints to stdout, so log cannot be used
+    local -r addon_path=$1
+    local -r obj_type=$2
+
+    for filePath in $(find $addon_path -name \*.yaml); do
+        # WARNING: assumption that the topmost indentation is zero (I'm not sure yaml allows for topmost indentation)
+        cat "${filePath}" | grep "^kind: ${obj_type}" >/dev/null 2>/dev/null 
+        if [[ $? -eq 0 ]]; then
+            echo $filePath
+        fi
+    done
+}
+
+# $1 file path
+function get-object-name-from-file() {
+    # prints to stdout, so log cannot be used
+    #WARNING: only yaml is supported
+    local -r name=$(cat $1 | python -c '''
```

Comment:
Split initialization of local and declaration to avoid errexit getting swallowed (bash gotcha). In this case, you can't use `local -r` either, because of this issue:

```
local name
name=$(blah)
```

If you do:
```
local -r name=$(foo)
```
the errexit can just get swallowed.